### PR TITLE
feat(core): feature flipping hook refacto

### PR DESCRIPTION
**Pull request by stif59100**
ref:MANAGER-13141

| Question         | Answer
| ---------------- | ---
| Branch?          | `feat/MANAGER-12621
| Bug fix?         | no
| New feature?     | yes
| Breaking change? | no
| Tickets          | MANAGER-13141
| License          | BSD 3-Clause

- [x] Try to keep pull requests small so they can be easily reviewed.
- [x] Commits are signed-off
- [ ] ~~Only FR translations have been updated~~
- [x] Branch is up-to-date with target branch
- [x] Lint has passed locally
- [x] Standalone app was ran and tested locally
- [x] Ticket reference is mentioned in linked commits (internal only)
- [ ] ~~Breaking change is mentioned in relevant commits~~

## Description

I did refactoring for the hook of feature flipping.

### CODE REVIEW

File: packages/manager/core/application/src/hooks/useFeatureAvailability.ts [hunk 1]
```diff
@@ -1,47 +1,46 @@
...
+import { useQuery } from '@tanstack/react-query';
 
-export const featureAvailabilityBaseQueryKey = 'feature-availability-';
-
-export type UseFeatureAvailabilityParams = {
-  parent: string;
-  featureList?: string[];
-};
+/**
+ * exemples
```
Comment (anooparveti) [inline packages/manager/core/application/src/hooks/useFeatureAvailability.ts:5]:
```suggestion
 * examples
```

File: packages/manager/core/application/src/hooks/useFeatureAvailability.ts [hunk 2]
```diff
@@ -1,47 +1,46 @@
...
-  const featuresListToCheck = featureList.flatMap(getFeaturesListToCheck);
 
+interface FeatureAvailabilityResult {
+  features: Record<string, boolean>;
+}
+
+export const useFeatureAvailability = (
+  featureList: string[],
+): {
+  data: FeatureAvailabilityResult | undefined;
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useFeatureAvailability.ts:23]:
```suggestion
  data?: FeatureAvailabilityResult;
```

So you give the result of feature availability call as data to the hook ?

Comment (chipp972) [inline packages/manager/core/application/src/hooks/useFeatureAvailability.ts:23]:
```suggestion
  data: FeatureAvailabilityResult;
```

Why not return directly the data ?

File: packages/manager/core/application/src/hooks/useFeatureAvailability.ts [hunk 3]
```diff
@@ -1,47 +1,46 @@
...
   const fetchFeatureAvailabilityData = async () => {
     const result = await apiClient.aapi.get(
-      `/feature/${featuresListToCheck.join(',')}/availability`,
+      `/feature/${featureList.join(',')}/availability`,
     );
-    return result.data || {};
+
+    const featureAvailabilityResult: FeatureAvailabilityResult = {
+      features: {},
+    };
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useFeatureAvailability.ts:32]: Why put it in an object ?

File: packages/manager/core/application/src/hooks/useFeatureAvailability.ts [hunk 4]
```diff
@@ -1,47 +1,46 @@
...
+    return featureAvailabilityResult;
   };
 
-  return useQuery<Record<string, boolean>>(
-    [`${featureAvailabilityBaseQueryKey}-${featuresListToCheck.join('-')}`],
+  return useQuery<FeatureAvailabilityResult>(
+    [featureList.join('-')],
     fetchFeatureAvailabilityData,
-  );
+  ) as { data: FeatureAvailabilityResult };
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useFeatureAvailability.ts:45]: Use the object syntax ({ queryKey: [featureList.join('-')], queryFn: fetchFeature... })

And don't change the type, useQuery already gives a better type

File: packages/manager/core/application/src/hooks/useFeatureAvailability.ts [hunk 5]
```diff
@@ -1,47 +1,40 @@
...
+      `/feature/${featureList.join(',')}/availability`,
     );
-    return result.data || {};
+
+    return {
+      features: featureList.reduce((acc, feature) => {
+        acc[feature] = result.data[feature] || false;
+        return acc;
+      }, {} as Record<string, boolean>),
+    };
```
Comment (anooparveti) [inline packages/manager/core/application/src/hooks/useFeatureAvailability.ts:33]: **question:** What's the issue with directly returning `result.data`?

File: packages/manager/core/application/src/hooks/useFeatureAvailability.ts [hunk 6]
```diff
@@ -1,47 +1,40 @@
-import { useQuery } from '@tanstack/react-query';
 import { apiClient } from '@ovh-ux/manager-core-api';
+import { useQuery } from '@tanstack/react-query';
 
-export const featureAvailabilityBaseQueryKey = 'feature-availability-';
+/**
+ * Examples
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useFeatureAvailability.ts:5]:
```suggestion
 * @example
```

File: packages/manager/core/application/src/hooks/useFeatureAvailability.ts [hunk 7]
```diff
@@ -1,47 +1,40 @@
...
+ * const isBillingAvailable = features.billing;
+ * const isWebooooAvailable = features.webooo;
+ * const isMicrosoftAvailable = features?.['web:microsoft'];
  */
-export const useFeatureAvailability = (
-  ...featureList: UseFeatureAvailabilityParams[]
-) => {
-  const featuresListToCheck = featureList.flatMap(getFeaturesListToCheck);
-
+const useFeatureAvailability = (
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useFeatureAvailability.ts:18]:
```suggestion
export const useFeatureAvailability = (
```

File: packages/manager/core/application/src/hooks/useFeatureAvailability.ts [hunk 8]
```diff
@@ -1,47 +1,39 @@
...
-};
-
-const getFeaturesListToCheck = ({
-  parent,
-  featureList = [],
-}: UseFeatureAvailabilityParams) =>
-  [parent].concat(featureList.map((name) => `${parent}:${name}`));
+export interface FeatureAvailabilityResult {
+  features: Record<string, boolean>;
+}
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useFeatureAvailability.ts:6]:
```suggestion
```

**Review by chipp972 — APPROVED** [review]

**Review by JacquesLarique — APPROVED** [review]

**Review by aboungnaseng-ovhcloud — APPROVED** [review]

**Pull request merged**